### PR TITLE
cluster-api: introduce prowjob-generator configuration and generate prowjobs

### PR DESCRIPTION
**Pull request by chrischdi**
Part of https://github.com/kubernetes-sigs/cluster-api/pull/9937

Introduces config + templates for prowjob-gen and generate configurations

### CODE REVIEW

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-prowjob-gen.yaml [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        to: "1.25"
+      - from: "1.25"
+        to: "1.26"
+      - from: "1.26"
+        to: "1.27"
+
+  # The generator will run each template per branch and use a filename friendly
+  # branch variable to format the target filename.
+  templates:
+  - name: "cluster-api-periodics.yaml.tpl"
```
Comment (sbueringer) [inline config/jobs/kubernetes-sigs/cluster-api/cluster-api-prowjob-gen.yaml:97]: (Please in a follow-up PR)

I thin I would change the output format.

Today the resulting grouping is a bit strange

image

Basically first all periodics then grouped by branch then normal + upgrade
Then all presubmits

I think I would put the branch at the beginning cluster-api-{branch}-periodics / cluster-api-{branch}-periodics-upgrades cluster-api-{branch}-presubmits

File: config/jobs/kubernetes-sigs/cluster-api/templates/cluster-api-periodics-upgrades.yaml.tpl [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      - name: ALWAYS_BUILD_KIND_IMAGES
+        value: "true"
+      - name: KUBERNETES_VERSION_UPGRADE_FROM
+        value: "{{ index (index $.versions $upgrade.From) "k8sRelease" }}"
+      - name: KUBERNETES_VERSION_UPGRADE_TO
+        value: "{{ index (index $.versions $upgrade.To) "k8sRelease" }}"
+      - name: ETCD_VERSION_UPGRADE_TO
+        value: "{{ index (index $.versions $upgrade.From) "etcd" }}"
+      - name: COREDNS_VERSION_UPGRADE_TO
+        value: "{{ index (index $.versions $upgrade.From) "coreDNS" }}"
```
Comment (sbueringer) [inline config/jobs/kubernetes-sigs/cluster-api/templates/cluster-api-periodics-upgrades.yaml.tpl:39]: This should be upgrade.To

Probably also means that the config has to be updated

File: config/jobs/kubernetes-sigs/cluster-api/templates/cluster-api-presubmits.yaml.tpl [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+          - name: ALWAYS_BUILD_KIND_IMAGES
+            value: "true"
+          - name: KUBERNETES_VERSION_UPGRADE_FROM
+            value: "{{ index (index $.versions ((last $.config.Upgrades).From)) "k8sRelease" }}"
+          - name: KUBERNETES_VERSION_UPGRADE_TO
+            value: "{{ index (index $.versions ((last $.config.Upgrades).To)) "k8sRelease" }}"
+          - name: ETCD_VERSION_UPGRADE_TO
+            value: "{{ index (index $.versions ((last $.config.Upgrades).From)) "etcd" }}"
+          - name: COREDNS_VERSION_UPGRADE_TO
+            value: "{{ index (index $.versions ((last $.config.Upgrades).From)) "coreDNS" }}"
```
Comment (sbueringer) [inline config/jobs/kubernetes-sigs/cluster-api/templates/cluster-api-presubmits.yaml.tpl:403]: Should be also .To

**Comment by sbueringer** [comment]
lgtm +/- the linter findings

I would merge as soon as the findings are addressed (I want to avoid getting churn through an auto-bump)

**Comment by sbueringer** [comment]
Thank you!

Merging for now. Let's open follow-up PRs if we want to change something (e.g. if the core CAPI PR changes)

As mentioned, I want to avoid a rebase if Prow gets bumped.

/lgtm
/approve

**Comment by chrischdi** [comment]
/hold

Should I squash?

**Comment by sbueringer** [comment]
No idea in this repo

Can you fix https://github.com/kubernetes/test-infra/pull/31581#discussion_r1452535995?

Otherwise the link won't work

**Comment by sbueringer** [comment]
Thx!

/lgtm
/approve

**Pull request merged**